scripts/util/math.ts

```diff
@@ -1,40 +1,36 @@
-/**
- * Utility functions for Javascript.
- * Could be ported to C++ and exposed globally in the future.
- */
+export function magnitude(vec: vec2 | vec3): number {
+	return Math.sqrt(sumOfSquares(vec));
+}
 
-/**
- * 2D vector object with x and y member variables
- */
-type Vec2D = { x: number; y: number };
+export function magnitude2D(vec: vec2): number {
+	return Math.sqrt(sumOfSquares2D(vec));
+}
 
-/**
- * 3D vector object with x, y, and z member variables
- */
-type Vec3D = { x: number; y: number; z: number };
-type Vector = Vec2D | Vec3D;
+export function magnitude3D(vec: vec3): number {
+	return Math.sqrt(sumOfSquares3D(vec));
+}
 
-/**
- * 2D length of input vector
- */
-function getSize2D(vec: Vector): number {
-	return Math.sqrt(getSizeSquared2D(vec));
+// Note: Math.hypot performs additional bounds checking on V8 which which makes it considerably
+// slower than below implementations.
+export function sumOfSquares2D(vec: vec2): number {
+	return vec.x ** 2 + vec.y ** 2;
 }
 
-/**
- * 2D length squared of input vector
- */
-function getSizeSquared2D(vec: Vector): number {
-	return vec.x * vec.x + vec.y * vec.y;
+export function sumOfSquares3D(vec: vec3): number {
+	return vec.x ** 2 + vec.y ** 2 + vec.z ** 2;
+}
+
+export function sumOfSquares(vec: vec2 | vec3): number {
+	return 'z' in vec ? vec.x ** 2 + vec.y ** 2 + vec.z ** 2 : vec.x ** 2 + vec.y ** 2;
```

Comment:
should this call the 2D or 3D version passing `vec` instead of redefining the operation?

Reply:
I guess so, functions calls in here probably get inlined so no performance benefit doing it this way.

Reply:
Spent 5 mins playing around with this, indeed there's no noticeable performance difference

Reply:
My bad, I wasn't super clear about this:
`return 'z' in vec ? sumOfSquares3D(vec) : sumOfSquares2D(vec);`